### PR TITLE
Merge MaDKitLanEdition with MaDKit ?

### PR DESCRIPTION
**Pull request by JasonMahdjoub**
Hi fab !

I have released MaDKitLanEdition 1.7.0 which is a very stable version. I asked myself if you are interessed to release an enhenced MaDKit version (MaDKit 6 ?). So I propose you to merge MKLE with MK. However, I am aware of the deep changements I have made into MKLE, and I can understand you have hesitations. Here, these changements :
* First, MKLE is an evolution of MaDKitGroupExtension. So groups are represented by a tree as it is the case with directories into file system. So an agent can request a role into a group which is itself contained into another group. See the class Group. More over, it is possible to represent several groups throuth the class MultiGroup. Groups classes have the functions include, equals, interset, union... I think the mangement of groups the more discusting point since there is no publication that make a theory of this kind of group's representation. However, this implementation does not remove anything from the AGR principle.
* A new kind of agent is created : AgentFakeThread. This agent have a function liveCycle which is called only if the agent reveive a message. On the other case, the agent does not use the thread nor the CPU. So you can create a high number of AgentFakeThread, they will take only 4 threads if your processor have 4 cores.
* In the classe Agent, the function live is renamed to liveCycle in order to represent a cycle of the agent and not its complete life.
* When killing an agent, it can take the state 'ZOMBIE' if its liveCycle function or end function never really end.
* MKLE support properties files, but also XML and YAML file configuration. YAML is the default format.
* The project is compiled with Gradle. I think it is more interesting than Ant since it is able to donwload project dependencies automatically, and since it is possible to include Groovy code to make the build more specific.
* Dependencies (including those writed by me) are in the server of DistriMind, however, I can move them into a more popular server like MavenCentral or JCenter.
* the package of MKLE is 'com.distrimind.madkit'. At the origin, I have though to make a clear separation between MK and MKLE. But If you consider to accept this pull request or an adapted pull request, I will rebase the principal package to 'com.madkit'.
* The most important changement with MKLE is obviously the lan part. 
    * MKLE manage direct connections between two peers, or indirect connections with one peer (or more) between two peers
    * A connection can be in clear or secured through an encryption/signature pipe. Intermediate peer does not see the messages in clear
    * P2P authentication is provided to enable the login between two peers without compromising their key/password, or their signature keys.
    * Serialization process is replaced by an externalization, and memory allocation is controlled in case of attack.
    * MKLE is conceived to limit DDOS attacks (maximum received message at the same time, maximum memory consumption for one connection and for all connections).
    * I really suggest you to read the README.md file into my repository.
* I have tried to generate documentation with the code, and this last is present when necessary. However, network connections lack of demonstration (only available into the junit tests). However, I have adapted the original MaDKit demos which are avaible with MKLE.

Feel free to ask me some questions.

Jason.

### CODE REVIEW

**Comment by fmichel** [comment]
Hi Jason !

I apologize for answering so lately to your request but I had hard times doing all my working duties lately...

What a tremendous work you have done ! Congratulations !

The "not deeply thought through" answer is that I cannot validate your request: 
(1) it is too much at a time ;)
(2) It would profoundly change the nature of how the core concepts of MaDKit are implemented. For instance, how you manage group recursively is very interesting but, as you mentioned, discussing theoretical foundations of this solution should be done first. More importantly, your solution on this matter looks really appealing to me and I definitely have to check it. Still, I would say that it corresponds to a superior modeling level of AGR so that it is already "something which is more than MaDKit". My point is that one crucial aspect of MaDKit is that concept and implementation are kept as simple as possible so that people can come up with their own solution, e.g. for this kind of modeling, as you have done yourself in fact. This is a crucial aspect for me.

Another answer is that I would really like to consider all these features separately. Obviously the LAN part would be beneficial for MaDKit whatever the context and it would be great to integrate it into the main branch.
That said, I guess that your solution is deeply linked to the other changes you have made, and thus not so easy to isolate...
In the same way, it would be very nice if the MaDKitGroupExtension was designed as a library that anyone could use as an additional library for MaDKit, separated from MaDKit itself. I guess that it raises the same kind of issues...
 
Anyway, what you have done could indeed serve as a kind of roadmap considering what MDK 6 should integrate and I should take a deep look into your changes and discuss them with you as soon as we will have the opportunity. 

To summarize my feeling, it is not a pull request that you have done, it is a "let us start a brand new project together" request ;) And I like the idea! Creating a new repo and merging all that is thus also an option...

Best,
Fabien

**Comment by JasonMahdjoub** [comment]
Hi Fabien,

I understand that there is too much modifications with MKLE. So threre is no problem. I think you can recognize my pull requests without signature as you recognized my emails in the past :)

Concerning the descriptions of features as extensions, by having experimenting it with MadKitGroupExtension, it produce bad internal code that can give random results in a perspective of incremental approach. I could alter MaDKit in order to adapt it to receive future extensions, and it would be a better choice, but I have done the choice to fork it in order to work speediest. The other things is that some features are dependant from others. For exemple, multi-thread network need the AgentFakeThread concept. Without that, you cannot develop a large scale system (over a fuzzy threshold, using lot of thread decrease the computer's performance). 

I understand your opinion about the need of a simple group management that respects AGR principle. As you have saying it, the new group management approch is deeply included into MKLE. To be honest, I am not really motivated to come back to a previous policy : this new approach does not remove anything from the older, and if you want to make simple groups, you have still this possibility. However, if you have a validated scientific model that concern this new approch of group's management, I agree to make it evolving.  For now, if you want to study how it works, you can check this demo (taken from the original MK demos) : [https://github.com/JazZ51/MaDKitLanEdition/tree/master/MaDKitLanEdition/MaDKitLanEditionDemos/src/main](https://github.com/JazZ51/MaDKitLanEdition/tree/master/MaDKitLanEdition/MaDKitLanEditionDemos/src/main)

Concerning the MDK6 roadmap, I can say that the network part of MKLE is very stable and thinked to be very secured with its own protocol. I have spent lot of times with a big determination to eliminate every bug and to obtain the last version (1.7.6). So if you want to redevelop each feature... it's possible, but good luck :) However, I am of course open to improve (or simplify) the design of these features into a common work.

Best,
Jason.
